The "to_url" function returning the proper result for absolute local paths.

mrudko opened the pull request:
This pull request introduces a slight modification to the return value of `to_url` function. This change is motivated by the fact that the existing configuration of the function cannot properly deal with the HPC cluster directory structure. Performing the fv3net-style segmented simulations on HPC cluster (Stellar) requires locating of the restart files within the output directory. The output directory is local absolute path rather then the path on the cloud. 

Refactored public API:
- `return protocol + "://" + path`: this return value of the `to_url` function replaces the existing `return protocol + "://" + path.lstrip("/")`

- [x] Tests added
The test was added to the `test_fsspec_cloud.py` file.
The predicate (`assert` statement) was modified in the `test_read_last_segment` function (`test_append.py`).
A new predicate (`assert` statement) was added to the `test_read_last_segment` function (`test_append.py`). It checks the existence of the supplied path.

Resolves #2202

spencerkclark reviewed (commented):
This is looking good -- thanks @mrudko.  

Nice job following our PR template.  A few minor suggestions regarding that:
- I suggest making the PR title more specific, perhaps something indicating that the change helps `to_url` return the proper result for absolute local paths.
- One nit regarding your description is that I think it should note that this specifically pertains to locating restart files for fv3net-style segmented runs where the output directory is a local absolute path rather than a path on the cloud (we have been doing segmented runs where the output path is in the cloud for a long time).  
- Finally, it would be great if you could add "Resolves #2203" somewhere in the description.  This links the PR back to the reported issue, and also tells GitHub to automatically close the issue once this PR is merged.

I addition, I didn't realize we had another [test](https://github.com/ai2cm/fv3net/blob/b8146cf7979901e7b78613a298b1973b48ca80a1/workflows/prognostic_c48_run/tests/segmented_run/test_append.py#L7-L14) that was impacted by this change within the prognostic run workflow (see the output from our CI [here](https://app.circleci.com/pipelines/github/ai2cm/fv3net/14346/workflows/e6ec3539-7c07-4f1c-9664-c55ec3a833ad/jobs/135059)):

```
=================================== FAILURES ===================================
____________________________ test_read_last_segment ____________________________

tmpdir = local('/tmp/pytest-of-root/pytest-0/test_read_last_segment0')

    def test_read_last_segment(tmpdir):
        date1 = "20160101.000000"
        date2 = "20160102.000000"
        arts = tmpdir.mkdir("artifacts")
        arts.mkdir(date1)
        arts.mkdir(date2)
        ans = read_last_segment(str(tmpdir))
>       assert f"file:/{str(tmpdir)}/artifacts/20160102.000000" == ans
E       AssertionError: assert 'file://tmp/p...160102.000000' == 'file:///tmp/p...160102.000000'
E         - file://tmp/pytest-of-root/pytest-0/test_read_last_segment0/artifacts/20160102.000000
E         + file:///tmp/pytest-of-root/pytest-0/test_read_last_segment0/artifacts/20160102.000000
E         ?      +

tests/segmented_run/test_append.py:14: AssertionError
```

This is actually a nice test, since it creates a mock set of artifacts directories from a segmented run, and if it had been written with an additional check it would have caught this issue before we ran into it.  What I mean here is if it checked for the existence of `ans` as a path in addition to validating its value.  It would be great if you could add this check in addition to updating the existing assert statement to be consistent with what `read_last_segment` now returns, so that the test passes.

The additional check could be something as simple as:
```python
fs = vcm.cloud.get_fs(ans)
assert fs.exists(ans)
```
This check would have failed prior to the change you made in this PR.

spencerkclark commented:
The issue building the `fv3fit_torch` image looks unrelated:
```
#18 [14/14] RUN pip3 install dgl-cu113 dglgo -f https://data.dgl.ai/wheels/repo.html
#18 sha256:e6625a465442fa9ddc44ede99f8e292d1c06e3f08c86e4bf94fb78e728ecdb5c
#18 1.225 Looking in links: https://data.dgl.ai/wheels/repo.html
#18 1.379 ERROR: Could not find a version that satisfies the requirement dgl-cu113 (from versions: none)
#18 1.380 ERROR: No matching distribution found for dgl-cu113
#18 ERROR: executor failed running [/bin/sh -c pip3 install dgl-cu113 dglgo -f https://data.dgl.ai/wheels/repo.html]: exit code: 1
```
This could be a transient issue, so we'll see if it resolves itself upon your next push, or maybe someone else on our team will need to look into installing that library from another source.

mrudko commented:
Thanks @spencerclark. I have changed the title and the description of the pull request accordingly.

I experimented a little bit with the `test_read_last_segment` function. After introducing additional check for path existence into the test and retaining the return statement of the `to_url` function unmodified, the test indeed fails.

```
tmpdir = local('/tmp/pytest-of-mr7417/pytest-17/test_read_last_segment0')

    def test_read_last_segment(tmpdir):
        date1 = "20160101.000000"
        date2 = "20160102.000000"
        arts = tmpdir.mkdir("artifacts")
        arts.mkdir(date1)
        arts.mkdir(date2)
        ans = read_last_segment(str(tmpdir))
    #    assert f"file:/{str(tmpdir)}/artifacts/20160102.000000" == ans
        fs = get_fs(ans)
>       assert fs.exists(ans)
E       AssertionError: assert False
E        +  where False = <bound method AbstractFileSystem.exists of <fsspec.implementations.local.LocalFileSystem object at 0x145debb76670>>('file://tmp/pytest-of-mr7417/pytest-17/test_read_last_segment0/artifacts/20160102.000000')
E        +    where <bound method AbstractFileSystem.exists of <fsspec.implementations.local.LocalFileSystem object at 0x145debb76670>> = <fsspec.implementations.local.LocalFileSystem object at 0x145debb76670>.exists

test_append.py:17: AssertionError
```
I have introduced all the suggested changes to the existing state of the test. The test fails for the previous version of the `to_url` function (`return protocol + "://" + path.lstrip("/")`), and it passes for the modified version of the `to_url` function (`return protocol + "://" + path`).

spencerkclark reviewed (approved):
Perfect @mrudko!  I'll likely go ahead and merge this later today.

oliverwm1 commented:
LGTM, I agree that this function should not strip the leading `/` of the provided path since it's clearly important to indicate that it's an absolute path. Although I admit I'm not totally clear on the behavior of the `file` protocol of `fsspec` when using relative paths.

spencerkclark commented:
Thanks @oliverwm1 -- in some experiments locally using a relative path with the `file` protocol seems to work as one might expect:
```
$ touch foo.txt
$ python
>>> import fsspec; import vcm
>>> fs = vcm.cloud.get_fs("foo.txt")
>>> fs.exists("file://foo.txt")
True
>>> fs.exists("file:///foo.txt")
False
```
